src/Compilers/CSharp/Test/Semantic/Semantics/RefFieldTests.cs

```diff
@@ -0,0 +1,2984 @@
+                Diagnostic(ErrorCode.ERR_FeatureInPreview, "ref readonly T").WithArguments("ref fields").WithLocation(3, 12));
+
+            verifyField(comp.GetMember<FieldSymbol>("S.F"), RefKind.RefReadOnly, "ref readonly T S<T>.F");
+
+            comp = CreateCompilation(sourceA);
+            comp.VerifyEmitDiagnostics();
+            var refA = AsReference(comp, useCompilationReference);
+            verifyField(comp.GetMember<FieldSymbol>("S.F"), RefKind.RefReadOnly, "ref readonly T S<T>.F");
+
+            var sourceB =
+@"using System;
+class A
+{
+    internal int G;
+}
+class Program
+{
+    static void Main()
+    {
+        A a = new A();
+        a.G = 1;
+        var s = new S<A>(in a);
+        s.F.G = 2;
+        Console.WriteLine(s.F.G);
+        Console.WriteLine(a.G);
+        a.G = 3;
+        Console.WriteLine(s.F.G);
+        Console.WriteLine(a.G);
+    }
+}";
+
+            comp = CreateCompilation(sourceB, references: new[] { refA }, parseOptions: TestOptions.Regular10);
+            comp.VerifyEmitDiagnostics(
+                // (13,9): error CS8652: The feature 'ref fields' is currently in Preview and *unsupported*. To use Preview features, use the 'preview' language version.
+                //         s.F.G = 2;
+                Diagnostic(ErrorCode.ERR_FeatureInPreview, "s.F").WithArguments("ref fields").WithLocation(13, 9),
+                // (14,27): error CS8652: The feature 'ref fields' is currently in Preview and *unsupported*. To use Preview features, use the 'preview' language version.
+                //         Console.WriteLine(s.F.G);
+                Diagnostic(ErrorCode.ERR_FeatureInPreview, "s.F").WithArguments("ref fields").WithLocation(14, 27),
+                // (17,27): error CS8652: The feature 'ref fields' is currently in Preview and *unsupported*. To use Preview features, use the 'preview' language version.
+                //         Console.WriteLine(s.F.G);
+                Diagnostic(ErrorCode.ERR_FeatureInPreview, "s.F").WithArguments("ref fields").WithLocation(17, 27));
+
+            verifyField(comp.GetMember<FieldSymbol>("S.F"), RefKind.RefReadOnly, "ref readonly T S<T>.F");
+
+            var verifier = CompileAndVerify(sourceB, references: new[] { refA }, parseOptions: TestOptions.RegularNext, verify: Verification.Skipped, expectedOutput: IncludeExpectedOutput(
+@"2
+2
+3
+3
+"));
+            comp = (CSharpCompilation)verifier.Compilation;
+            verifyField(comp.GetMember<FieldSymbol>("S.F"), RefKind.RefReadOnly, "ref readonly T S<T>.F");
+
+            static void verifyField(FieldSymbol field, RefKind refKind, string displayName)
+            {
+                Assert.Equal(refKind, field.RefKind);
+                Assert.Equal(displayName, field.ToTestDisplayString());
+            }
+        }
+
+        [Fact]
+        public void SubstitutedField()
+        {
+            var sourceA =
+@".class public A<T>
+{
+  .field public !0& modopt(object) modopt(int8) F
```

Comment:
What does this series of `modopt` produce here? I'm clear how this is related to `ref` fields

Reply:
The `modopt` are the `RefCustomModifiers` of the field.

Reply:
Ah I missed the trailing `&` there that turned it into a `ref`